### PR TITLE
kubelet: don't reject pods without adding them to the pod manager

### LINKED ISSUE #37658: Pods rejected by Kubelet might stay at pending phase forever
Forked from https://github.com/kubernetes/kubernetes/issues/37447#issuecomment-263719696:

@yujuhong wrote:

For (2) in my previous comment, I've verified that this is a kubelet bug (i.e., not related to the apiserver watch). What I did was:

Edit the image used in the pending pod (kubectl edit pod)
Get a UPDATE message from the kubelet log:
Nov 29 21:44:14 gke-test-cluster-1-default-pool-44d8ae2f-kpn9 kubelet[2228]: I1129 21:44:14.988756    2228 kubelet.go:2300] SyncLoop (UPDATE, "api"): "my-app-2781155866-wqtqt_default(72d911fd-b1e4-11e6-848f-42010a840181)"
What this means is that kubelet itself thinks this is not a new pod, but merely an update. To further confirm this, I access kubelet's /pods endpoint:

            "metadata": {
                "annotations": {
                    "kubernetes.io/config.seen": "2016-11-24T01:24:05.362036637Z",
                    "kubernetes.io/config.source": "api",
                    "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicaSet\",\"namespace\":\"default\",\"name\":\"my-app-2781155866\",\"uid\":\"48
22986e-b181-11e6-b21f-42010a840181\",\"apiVersion\":\"extensions\",\"resourceVersion\":\"182642\"}}\n"
                },
                "creationTimestamp": "2016-11-24T01:22:21Z",
                "generateName": "my-app-2781155866-",
                "labels": {
                    "name": "my-app",
                    "pod-template-hash": "2781155866"
                },
                "name": "my-app-2781155866-wqtqt",
                "namespace": "default",
                "ownerReferences": [
                    {
                        "apiVersion": "extensions/v1beta1",
                        "controller": true,
                        "kind": "ReplicaSet",
                        "name": "my-app-2781155866",
                        "uid": "4822986e-b181-11e6-b21f-42010a840181"
                    }
                ],
                "resourceVersion": "182846",
                "selfLink": "/api/v1/namespaces/default/pods/my-app-2781155866-wqtqt",
                "uid": "72d911fd-b1e4-11e6-848f-42010a840181"
kubelet marked the pod as seen on 11/24: kubernetes.io/config.seen": "2016-11-24T01:24:05.362036637Z"

There are only a few places where this update could've been dropped by kubelet. I checked the code and found a regression that caused by a change in #28570. Will send a fix later.

### PR DESCRIPTION
**Pull request by yujuhong**
kubelet relies on the pod manager as a cache of the pods in the apiserver (and
other sources) . The cache should be kept up-to-date even when rejecting pods.
Without this, kubelet may decide at any point to drop the status update
(request to the apiserver) for the rejected pod since it would think the pod no
longer exists in the apiserver.

This should fix #37658

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37661)

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1911,22 +1911,24 @@ func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
 	start := kl.clock.Now()
 	sort.Sort(sliceutils.PodsByCreationTime(pods))
 	for _, pod := range pods {
+		existingPods := kl.podManager.GetPods()
+		// Always add the pod to the pod manager. Kubelet relies on the pod
+		// manager as the source of truth for the desired state. If a pod does
+		// not esist in the pod manager, it means that it has been deleted in
```
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: s/esist/exist

**Comment by Random-Liu** [comment]
LGTM. Thanks for the fix. And sorry for not catching this when review that PR. :p

**Comment by derekwaynecarr** [comment]
@sjenning - I think this may also fix the issue we saw today where pods that were stuck terminating even when containers were all terminated remained stuck after a Kubelet restart.  It seems those pods were never getting their status updated which would happen if they had been filtered out of pod manager.   Need to verify this more though .

**Comment by dims** [comment]
@k8s-bot gci gke e2e test this

**Comment by calebamiles** [comment]
should this get a cherry pick label @dchen1107, @Random-Liu?

**cc: @saad-ali, @dims, @kubernetes/sig-node**

**Comment by yujuhong** [comment]
> should this get a cherry pick label

Yes. Just added.

**Comment by yujuhong** [comment]
FWIW, I reproduced this issue in my own cluster using a somewhat extreme test case (>1000 pods assigned to one node), and verified the fix. However, it's good to keep in mind that the kubelet's apiserver client has limited QPS and if the replication control keeps creating new pods assigned to the node, kubelet's status update throughout would never be able to catch up.

**Comment by yujuhong** [comment]
@dchen1107 @Random-Liu I added a check to see whether the pod in question has terminated or not. PTAL again, thanks!

/cc @dashpole

**Comment by yujuhong** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by Random-Liu** [comment]
LGTM

**Pull request merged**